File: backend/src/ontoforge_server/core/embedding.py
import logging
from abc import ABC, abstractmethod

import httpx

from ontoforge_server.config import settings

logger = logging.getLogger(__name__)

_provider: "EmbeddingProvider | None" = None
_client: httpx.AsyncClient | None = None

_DEFAULT_DIMENSIONS = {
    "ollama": 768,
    "openai": 1536,
}


class EmbeddingProvider(ABC):
    @abstractmethod
    async def embed(self, text: str) -> list[float] | None:
        """Generate an embedding vector for the given text.

        Returns None on error (caller proceeds without embedding).
        """

    @property
    @abstractmethod
    def dimensions(self) -> int:
        """Return the dimensionality of the embedding vectors."""


class OllamaEmbeddingProvider(EmbeddingProvider):
    def __init__(
        self, model: str, base_url: str, dims: int, client: httpx.AsyncClient,
    ) -> None:
        self._model = model
        self._base_url = base_url.rstrip("/")
        self._dims = dims
        self._client = client

    async def embed(self, text: str) -> list[float] | None:
        try:
            response = await self._client.post(
                f"{self._base_url}/api/embeddings",
                json={"model": self._model, "prompt": text},
            )
            response.raise_for_status()
            return response.json()["embedding"]
        except Exception as exc:
            logger.warning("Embedding failed: %s", exc)
            return None

    @property
    def dimensions(self) -> int:
        return self._dims


class OpenAIEmbeddingProvider(EmbeddingProvider):
    """OpenAI-compatible embedding provider.

    Works with any API that follows the OpenAI embeddings format
    (OpenAI, Azure OpenAI, vLLM, LM Studio, etc.).
    """

    def __init__(
        self,
        model: str,
        base_url: str,
        api_key: str,
        dims: int,
        client: httpx.AsyncClient,
    ) -> None:
        self._model = model
        self._base_url = base_url.rstrip("/")
        self._api_key = api_key
        self._dims = dims
        self._client = client

    async def embed(self, text: str) -> list[float] | None:
        try:
            response = await self._client.post(
                f"{self._base_url}/v1/embeddings",
                headers={"Authorization": f"Bearer {self._api_key}"},
                json={"input": text, "model": self._model},
            )
            response.raise_for_status()
            return response.json()["data"][0]["embedding"]
        except Exception as exc:
            logger.warning("Embedding failed: %s", exc)
            return None

    @property
    def dimensions(self) -> int:
        return self._dims


def _resolve_dimensions(provider: str) -> int:
    if settings.EMBEDDING_DIMENSIONS is not None:
        return settings.EMBEDDING_DIMENSIONS
    default = _DEFAULT_DIMENSIONS.get(provider)
    if default is not None:
        return default
    raise ValueError(
        f"EMBEDDING_DIMENSIONS is required for provider '{provider}'"
    )


def create_embedding_provider(
    provider: str, model: str, base_url: str, client: httpx.AsyncClient,
) -> EmbeddingProvider:
    dims = _resolve_dimensions(provider)
    if provider == "ollama":
        return OllamaEmbeddingProvider(model, base_url, dims, client)
    if provider == "openai":
        api_key = settings.EMBEDDING_API_KEY
        if not api_key:
            raise ValueError("EMBEDDING_API_KEY is required for the openai provider")
        return OpenAIEmbeddingProvider(model, base_url, api_key, dims, client)
    raise ValueError(f"Unknown embedding provider: '{provider}'")


async def init_embedding_provider() -> None:
    global _provider, _client
    if not settings.EMBEDDING_PROVIDER:
        logger.info("EMBEDDING_PROVIDER not set — semantic search disabled")
        return
    _client = httpx.AsyncClient(timeout=30.0)
    _provider = create_embedding_provider(
        settings.EMBEDDING_PROVIDER,
        settings.EMBEDDING_MODEL,
        settings.EMBEDDING_BASE_URL,
        _client,
    )
    logger.info(
        "Embedding provider initialized: %s (%s, %d dimensions)",
        settings.EMBEDDING_PROVIDER,
        settings.EMBEDDING_MODEL,
        _provider.dimensions,
    )


async def close_embedding_provider() -> None:
    global _provider, _client
    if _client is not None:
        await _client.aclose()
        _client = None
    _provider = None


def get_embedding_provider() -> "EmbeddingProvider | None":
    return _provider

File: server/src/core/embedding.ts
/**
 * Embedding provider seam and implementations (`core/embedding` in the
 * module layout).
 *
 * Two providers: `ollama` (native embeddings endpoint) and `openai`
 * (OpenAI-compatible `/v1/embeddings` — works with OpenAI, Azure, vLLM,
 * LM Studio, …). A failed embedding call is LOGGED and yields `null`; the
 * caller proceeds without a vector — an embedding failure never fails a
 * write (`docs/capabilities/search.md#keeping-embeddings-current`).
 *
 * With no `EMBEDDING_PROVIDER` configured, no provider is ever installed
 * and every consumer that gates on `getEmbeddingProvider()` — chunk
 * synchronization, entity embedding, semantic search, vector-index DDL —
 * is a no-op. Tests inject a fake provider to exercise the gated paths.
 */

import { settings } from "../config.js";

export interface EmbeddingProvider {
  /** Vector width, needed for index DDL. */
  readonly dimensions: number;
  /** Embed one text. `null` means the provider produced no vector. */
  embed(text: string): Promise<number[] | null>;
}

/** `fetch`-shaped dependency so unit tests can inject a fake transport. */
export type FetchFn = (input: string, init?: RequestInit) => Promise<Response>;

const REQUEST_TIMEOUT_MS = 30_000;

const DEFAULT_DIMENSIONS: Record<string, number> = {
  ollama: 768,
  openai: 1536,
};

export class OllamaEmbeddingProvider implements EmbeddingProvider {
  private readonly baseUrl: string;

  constructor(
    private readonly model: string,
    baseUrl: string,
    readonly dimensions: number,
    private readonly fetchFn: FetchFn = fetch,
  ) {
    this.baseUrl = baseUrl.replace(/\/+$/, "");
  }

  async embed(text: string): Promise<number[] | null> {
    try {
      const response = await this.fetchFn(`${this.baseUrl}/api/embeddings`, {
        method: "POST",
        headers: { "content-type": "application/json" },
        body: JSON.stringify({ model: this.model, prompt: text }),
        signal: AbortSignal.timeout(REQUEST_TIMEOUT_MS),
      });
      if (!response.ok) {
        throw new Error(`HTTP ${response.status}`);
      }
      const payload = (await response.json()) as { embedding: number[] };
      return payload.embedding;
    } catch (exc) {
      console.warn(`Embedding failed: ${exc instanceof Error ? exc.message : String(exc)}`);
      return null;
    }
  }
}

export class OpenAIEmbeddingProvider implements EmbeddingProvider {
  private readonly baseUrl: string;

  constructor(
    private readonly model: string,
    baseUrl: string,
    private readonly apiKey: string,
    readonly dimensions: number,
    private readonly fetchFn: FetchFn = fetch,
  ) {
    this.baseUrl = baseUrl.replace(/\/+$/, "");
  }

  async embed(text: string): Promise<number[] | null> {
    try {
      const response = await this.fetchFn(`${this.baseUrl}/v1/embeddings`, {
        method: "POST",
        headers: {
          "content-type": "application/json",
          authorization: `Bearer ${this.apiKey}`,
        },
        body: JSON.stringify({ input: text, model: this.model }),
        signal: AbortSignal.timeout(REQUEST_TIMEOUT_MS),
      });
      if (!response.ok) {
        throw new Error(`HTTP ${response.status}`);
      }
      const payload = (await response.json()) as { data: { embedding: number[] }[] };
      return payload.data[0]!.embedding;
    } catch (exc) {
      console.warn(`Embedding failed: ${exc instanceof Error ? exc.message : String(exc)}`);
      return null;
    }
  }
}

function resolveDimensions(provider: string): number {
  if (settings.EMBEDDING_DIMENSIONS !== null) {
    return settings.EMBEDDING_DIMENSIONS;
  }
  const fallback = DEFAULT_DIMENSIONS[provider];
  if (fallback !== undefined) {
    return fallback;
  }
  throw new Error(`EMBEDDING_DIMENSIONS is required for provider '${provider}'`);
}

/** Build a provider from its name; throws on unknown names and missing
 * credentials — startup fails loudly rather than serving degraded. */
export function createEmbeddingProvider(
  provider: string,
  model: string,
  baseUrl: string,
  fetchFn: FetchFn = fetch,
): EmbeddingProvider {
  const dims = resolveDimensions(provider);
  if (provider === "ollama") {
    return new OllamaEmbeddingProvider(model, baseUrl, dims, fetchFn);
  }
  if (provider === "openai") {
    const apiKey = settings.EMBEDDING_API_KEY;
    if (!apiKey) {
      throw new Error("EMBEDDING_API_KEY is required for the openai provider");
    }
    return new OpenAIEmbeddingProvider(model, baseUrl, apiKey, dims, fetchFn);
  }
  throw new Error(`Unknown embedding provider: '${provider}'`);
}

let provider: EmbeddingProvider | null = null;

/** Startup step 3: install the configured provider, or none. */
export function initEmbeddingProvider(): void {
  if (!settings.EMBEDDING_PROVIDER) {
    console.info("EMBEDDING_PROVIDER not set — semantic search disabled");
    return;
  }
  provider = createEmbeddingProvider(
    settings.EMBEDDING_PROVIDER,
    settings.EMBEDDING_MODEL,
    settings.EMBEDDING_BASE_URL,
  );
  console.info(
    `Embedding provider initialized: ${settings.EMBEDDING_PROVIDER} ` +
      `(${settings.EMBEDDING_MODEL}, ${provider.dimensions} dimensions)`,
  );
}

export function closeEmbeddingProvider(): void {
  provider = null;
}

/** The active provider, or `null` when embeddings are disabled. */
export function getEmbeddingProvider(): EmbeddingProvider | null {
  return provider;
}

/** Install (or clear) the active provider. Startup and tests only. */
export function setEmbeddingProvider(next: EmbeddingProvider | null): void {
  provider = next;
}
